src/node_jsrtapi.cc

```diff
@@ -1004,67 +1047,157 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
 }
 
 napi_status napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
-  napi_destruct destructor, napi_weakref* handle) {
-  new ObjectWrapWrapper(jsObject, nativeObj, destructor);
+    napi_finalize finalize_cb, napi_ref* result) {
+  JsValueRef value = reinterpret_cast<JsValueRef>(jsObject);
```

Comment:
Yes. Normally it would be constructed via a NAPI function callback. If not, the call to `JsSetExternalData`, and therefore this function, will return an invalid-arg error.

I don't think it's worth representing it in the type system. It is sufficient to document that `napi_wrap` may only be used on the `this` parameter of a constructor callback function.